Add AudioDecoderCocoa class, a CoreMedia implementation of WebCodec's AudioEncoder

jyavenard opened the pull request:
#### 1e7fe684d4e5af6b98e6f15579db75334ed5d019
<pre>
Add AudioDecoderCocoa class, a CoreMedia implementation of WebCodec&apos;s AudioEncoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=283903">https://bugs.webkit.org/show_bug.cgi?id=283903</a>
<a href="https://rdar.apple.com/problem/140782290">rdar://problem/140782290</a>

Reviewed by Youenn Fablet.

We add the CoreMedia version of AudioDecoder which uses AudioSampleBufferConverter to decode any supported formats.
For now, the only thing we do not support are:
- vorbis if no description is provided in the AudioConfig object
- opus greater than 2 channels (this is a CoreMedia restrictions)

Covered by existing tests after updating expectations.

* LayoutTests/TestExpectations: Remove all AudioDecoder&apos;s related failure expectations. They all pass now.
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt: Tests full pass now; update expectations
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt: Tests full pass now; update expectations
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/transfering.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/transfering.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/mac/TestExpectations: on Ventura, Opus and MP3 decoder creation fails.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::isValidDecoderConfig):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported): Check that sample rate and channel count isn&apos;t 0.
(WebCore::AudioDecoder::create):
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp: Added.
(WebCore::queueSingleton):
(WebCore::InternalAudioDecoderCocoa::create):
(WebCore::InternalAudioDecoderCocoa::isConfigured const):
(WebCore::InternalAudioDecoderCocoa::converter):
(WebCore::AudioDecoderCocoa::create):
(WebCore::AudioDecoderCocoa::AudioDecoderCocoa):
(WebCore::AudioDecoderCocoa::isCodecSupported):
(WebCore::AudioDecoderCocoa::decode):
(WebCore::AudioDecoderCocoa::flush):
(WebCore::AudioDecoderCocoa::reset):
(WebCore::AudioDecoderCocoa::close):
(WebCore::InternalAudioDecoderCocoa::decompressedAudioOutputBufferCallback):
(WebCore::InternalAudioDecoderCocoa::processedDecodedOutputs):
(WebCore::InternalAudioDecoderCocoa::InternalAudioDecoderCocoa):
(WebCore::InternalAudioDecoderCocoa::initialize):
(WebCore::InternalAudioDecoderCocoa::stripADTSHeader):
(WebCore::InternalAudioDecoderCocoa::decode):
(WebCore::InternalAudioDecoderCocoa::flush):
(WebCore::InternalAudioDecoderCocoa::close):
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.h: Added.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h: Add default member initialiser so that we don&apos;t have to explicitly set them on construction.
* Source/WebCore/platform/mediastream/RealtimeOutgoingAudioSource.h:

Canonical link: <a href="https://commits.webkit.org/287440@main">https://commits.webkit.org/287440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b082a0c65901e1b9504b7ca3a577ac1bbff44e8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84268 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30751 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81845 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67810 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7024 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62323 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20162 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82805 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52376 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42630 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49717 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/26763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29189 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27228 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85691 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6974 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4864 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68457 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69820 "Found 4 new API test failures: /TestWebKit:WebKit.EnumerateDevicesCrash, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/test-page-list, /TestWebKit:WebKit.FirstMeaningfulPaint, /TestWebKit:WebKit2UserMessageRoundTripTest.WKURL (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17401 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13830 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12740 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6928 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8600 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -70,8 +70,16 @@ static bool isValidDecoderConfig(const WebCodecsAudioDecoderConfig& config)
+        return false;
+#endif
+
+    if (config.codec == "optus"_s && config.numberOfChannels > 2 && !config.description)
```
This is a typo i guess? :)

jyavenard replied:
oh :)

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -70,8 +70,16 @@ static bool isValidDecoderConfig(const WebCodecsAudioDecoderConfig& config)
+        return false;
+#endif
+
+    if (config.codec == "opus"_s && config.numberOfChannels > 2 && !config.description)
```
I am not sure where this check comes from, can you clarify?

jyavenard replied:
> I am not sure where this check comes from, can you clarify?

(https://github.com/w3c/webcodecs/issues/826)

other UA are doing it now

jyavenard replied:
You can't do Opus with more than 2 channels if you don't have the Ogg configuration anyway. So it's fine to reject that early. No UA can handle it

jyavenard replied:
I've removed for now.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.h:
```diff
@@ -0,0 +1,62 @@
+    void reset() final;
+    void close() final;
+
+    Ref<InternalAudioDecoderCocoa> m_internalDecoder;
```
const Ref so we do not have to take a Ref?

jyavenard replied:
is this enabled yet?

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+Ref<AudioDecoder::CreatePromise> AudioDecoderCocoa::create(const String& codecName, const Config& config, OutputCallback&& outputCallback)
+{
+    auto decoder = adoptRef(*new AudioDecoderCocoa(WTFMove(outputCallback)));
+    auto internalDecoder = decoder->m_internalDecoder;
```
Ref

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+
+Ref<AudioDecoder::CreatePromise> AudioDecoderCocoa::create(const String& codecName, const Config& config, OutputCallback&& outputCallback)
+{
+    auto decoder = adoptRef(*new AudioDecoderCocoa(WTFMove(outputCallback)));
```
Ref

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    if (codecName.startsWith("mp4a"_s)) {
+        if (codecName.startsWith("mp4a.40.2"_s))
+            codec = kAudioFormatMPEG4AAC;
```
It might read better if we would return early here and below.

jyavenard replied:
kAudioFormatMPEG4AAC is a CFENUM() it doesn't work well with `return makePair`

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    Ref<SharedBuffer> stripADTSHeader(SharedBuffer&);
+
+    const bool m_configured { false };
+    bool m_isClosed { false };
```
std::atomic.

jyavenard replied:
no, it's only accessed on the queue.

youennf replied:
m_isClosed is set to true in the WebCodecsAudioDecoder thread but read from the queue AIUI.

jyavenard replied:
it is not. WebCodecsAudioDecoder calls AudioDecoderCocoa::close() which dispatch a task on the queue to  call InternalAudioDecoderCocoa::close() which then set m_isClosed and then finalise the converter and then exit.

I'll added some GUARDED_BY_CAPABILITY to help show what's going on

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    Ref<SharedBuffer> stripADTSHeader(SharedBuffer&);
+
+    const bool m_configured { false };
```
Can we remove this one?

jyavenard replied:
yes.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    } else
+        m_inputDescription = CAAudioStreamDescription { double(config.sampleRate), uint32_t(config.numberOfChannels), *format, CAAudioStreamDescription::IsInterleaved::Yes };
+
+    m_inputFormatDescription = createAudioFormatDescription(*m_inputDescription, m_codecDescription.span());
```
m_codecDescription is never used but here.
Should we remove it from now and introduce it when we need it.

jyavenard replied:
actually, it's a mistake to not use it, it should be first set to config.description

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -70,8 +70,16 @@ static bool isValidDecoderConfig(const WebCodecsAudioDecoderConfig& config)
+
+#if USE(AVFOUNDATION)
+    if (config.codec == "opus"_s && config.numberOfChannels > 2)
+        return false;
```
This is not as per spec. Should we remove it and let the decoder creation fail in AV specific code instead?

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    void processedDecodedOutputs();
+
+    AudioDecoder::OutputCallback m_outputCallback;
+    RefPtr<AudioSampleBufferConverter> m_converter;
```
Maybe we should have a static function to create the converter and pass it to InternalAudioDecoderCocoa constructor. Then we could make it a `const Ref`.

jyavenard replied:
too complicated due to the arguments its creation requires. I'll give it a try

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    bool isConfigured() const { return !!m_converter; }
+
+private:
+    InternalAudioDecoderCocoa(AudioDecoder::OutputCallback&&);
```
explicit

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(AudioDecoderCocoa);
+
+static WorkQueue& queueSingleton()
```
Do we need this queue since `AudioSampleBufferConverter` is doing the work in a queue itself?

jyavenard replied:
it does make it easier to control as we control the lifetime of the runnable that way. If we were running on a worker, this is no longer guaranteed and that is an issue with the way the AudioSampleConverter returns the frame to us via a callback.

jyavenard replied:
As discussed offline, there's no additional copy required over running in the worker or main thread. We copy the incoming data in a SharedBuffer that we pass to the queue. We then create a CMBlockBuffer frame that is using the SharedBuffer as backend.

jyavenard replied:
On further thinking I think that at this stage, yes we do.
The AudioSampleBufferConverter uses both NativePromise and a callback on its queue.

To handle the NativePromise we need to know where the result should be dispatched. Having a dedicated queue is the simplest approach.

Otherwise, we would need to pass the WorkerThread to the AudioDecoderCocoa and use enqueueWhenSettled on the context.

2) the callback as it is managed only take a raw pointer, so making sure that pointer is valid to the end makes it not really suitable with a WorkerThread which could be deleted at any given time. We would need to dispatch a task regardless on the converter to finish it.

So I can look at this in a follow up task, but considering all {Audio|Video){Encoder|Decoder) follow the exact same patterns on using their own queue, I'm not sure why here we would do anything different particularly when it would make the code harder, not easier IMHO.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+void InternalAudioDecoderCocoa::decompressedAudioOutputBufferCallback(void* object, CMBufferQueueTriggerToken)
+{
+    // We can only be called from the CoreMedia callback if we are still alive.
+    RefPtr decoder = static_cast<class InternalAudioDecoderCocoa*>(object);
```
This is not totally clear how this is guaranteed here.
Can you clarify this?

For instance, `addSampleBuffer` is asynchronous and can run for some time.
What happens if the `AudioDecoderCocoa` is destroyed without close being called?

Maybe there should be a ref in addSampleBuffer lambda, or there is a need to do something in `AudioDecoderCocoa::~AudioDecoderCocoa()`?

jyavenard replied:
in close; we call finish() and hold a strong reference to ourselves until this is complete.

There can be no more call to the callback after that.

close() is guaranteed to be called by contract of the AudioDecoder and the WebCodecsAudioDecoder

youennf replied:
I do not think there is such a guarantee, say in case of stopActiveDOMObjects.
In any case, I think `AudioDecoder/InternalAudioDecoderCocoa` should be self protective, otherwise we might change slightly things in WebCodecsAudioDecoder and break the assumption that `InternalAudioDecoderCocoa` is making.

jyavenard replied:
I've made it more explicit. Calling close() in the destructor which holds a reference to itself until we're guaranteed that the converter won't output new packets

youennf replied:
Maybe add a comment in the destructor?

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    RefPtr decoder = static_cast<class InternalAudioDecoderCocoa*>(object);
+
+    queueSingleton().dispatch([weakDecoder = ThreadSafeWeakPtr { *decoder }] {
+        if (auto strongDecoder = weakDecoder.get())
```
RefPtr

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+        return AudioDecoder::DecodePromise::createAndReject("Decoder is closed"_s);
+
+    if (mIsAAC)
+        frameData = stripADTSHeader(frameData);
```
Do we need this additional allocation?
Can we create a span from frameData and use the span to create the block buffer?

jyavenard replied:
there is no additional allocation; what is returned is a SharedBufferDataView of the original SharedBuffer.

youennf replied:
`getContiguousData` is creating a new ref counted SharedBuffer object so there is a most probably an allocation, even if there is no copy of the underlying data.
I guess either way is fine if we would keep the new SharedBuffer around for some time, but here, we directly discard it after creating a block buffer.

jyavenard replied:
> getContiguousData is creating a new ref counted SharedBuffer object

it doesn't because the data is contiguous to start and so getContiguousData() returns itself.

jyavenard replied:
> I guess either way is fine if we would keep the new SharedBuffer around for some time, but here, we directly discard it after creating a block buffer.

we do not discard the SharedBuffer

We do 
```
RetainPtr blockBuffer = frameData->createCMBlockBuffer();
```

this create a blockBuffer that uses the original SharedBuffer as a reference, there's no copy involved.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,349 @@
+    LOG(Media, "Decoding%s frame", isKeyFrame ? " key" : "");
+
+    if (m_isClosed)
+        return AudioDecoder::DecodePromise::createAndReject("Decoder is closed"_s);
```
How about `using DecodePromise = AudioDecoder::DecodePromise;` to shorten these lines?

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+    if (components.size() != 2)
+        return makeUnexpected(makeString("Invalid LPCM codec string:"_s, codecName));
+    auto pcmFormat = components[1].convertToASCIILowercase();
+    codec = kAudioFormatLinearPCM;
```
Not needed

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+
+void AudioDecoderCocoa::reset()
+{
+    // FIXME: that can't be correct
```
The goal of reset to set closed synchronously was to limit the number of decode callback calls.
It should not be observable to dispatch to the work queue.

The FIXME could probably be removed.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+    if (!converter)
+        return;
+    // We keep a reference to ourselves until the converter has been marked as finished. This guarantees that no
+    // callback will occur after we are .
```
```suggestion
    // callback will occur after we are.
```

jyavenard committed:
Add AudioDecoderCocoa class, a CoreMedia implementation of WebCodec's AudioEncoder
https://bugs.webkit.org/show_bug.cgi?id=283903
rdar://problem/140782290

Reviewed by Youenn Fablet.

We add the CoreMedia version of AudioDecoder which uses AudioSampleBufferConverter to decode any supported formats.
For now, the only thing we do not support are:
- vorbis if no description is provided in the AudioConfig object
- opus greater than 2 channels (this is a CoreMedia restrictions)

Covered by existing tests after updating expectations.

* LayoutTests/TestExpectations: Remove all AudioDecoder's related failure expectations. They all pass now.
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt: Tests full pass now; update expectations
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt: Tests full pass now; update expectations
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/transfering.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/transfering.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/mac/TestExpectations: on Ventura, Opus and MP3 decoder creation fails.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::isValidDecoderConfig):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported): Check that sample rate and channel count isn't 0.
(WebCore::AudioDecoder::create):
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp: Added.
(WebCore::queueSingleton):
(WebCore::InternalAudioDecoderCocoa::create):
(WebCore::InternalAudioDecoderCocoa::isConfigured const):
(WebCore::InternalAudioDecoderCocoa::converter):
(WebCore::AudioDecoderCocoa::create):
(WebCore::AudioDecoderCocoa::AudioDecoderCocoa):
(WebCore::AudioDecoderCocoa::isCodecSupported):
(WebCore::AudioDecoderCocoa::decode):
(WebCore::AudioDecoderCocoa::flush):
(WebCore::AudioDecoderCocoa::reset):
(WebCore::AudioDecoderCocoa::close):
(WebCore::InternalAudioDecoderCocoa::decompressedAudioOutputBufferCallback):
(WebCore::InternalAudioDecoderCocoa::processedDecodedOutputs):
(WebCore::InternalAudioDecoderCocoa::InternalAudioDecoderCocoa):
(WebCore::InternalAudioDecoderCocoa::initialize):
(WebCore::InternalAudioDecoderCocoa::stripADTSHeader):
(WebCore::InternalAudioDecoderCocoa::decode):
(WebCore::InternalAudioDecoderCocoa::flush):
(WebCore::InternalAudioDecoderCocoa::close):
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.h: Added.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h: Add default member initialiser so that we don't have to explicitly set them on construction.
* Source/WebCore/platform/mediastream/RealtimeOutgoingAudioSource.h:

Canonical link: https://commits.webkit.org/287440@main
